pcsx2/Ps2Float.cpp

```diff
@@ -0,0 +1,791 @@
+
+int32_t Ps2Float::CompareTo(Ps2Float other)
+{
+	int32_t selfTwoComplementVal = (int32_t)(Abs());
+	if (Sign)
+		selfTwoComplementVal = -selfTwoComplementVal;
+
+	int32_t otherTwoComplementVal = (int32_t)(other.Abs());
+	if (other.Sign)
+		otherTwoComplementVal = -otherTwoComplementVal;
+
+	if (selfTwoComplementVal < otherTwoComplementVal)
+		return -1;
+	else if (selfTwoComplementVal == otherTwoComplementVal)
+		return 0;
+	else
+		return 1;
+}
+
+int32_t Ps2Float::CompareOperand(Ps2Float other)
+{
+	int32_t selfTwoComplementVal = (int32_t)(Abs());
+	int32_t otherTwoComplementVal = (int32_t)(other.Abs());
+
+	if (selfTwoComplementVal < otherTwoComplementVal)
+		return -1;
+	else if (selfTwoComplementVal == otherTwoComplementVal)
+		return 0;
+	else
+		return 1;
+}
+
+double Ps2Float::ToDouble()
+{
+	return std::bit_cast<double>(((u64)Sign << 63) | ((((u64)Exponent - BIAS) + 1023ULL) << 52) | ((u64)Mantissa << 29));
+}
+
+std::string Ps2Float::ToString()
+{
+	double res = ToDouble();
+
+	uint32_t value = AsUInt32();
+	std::ostringstream oss;
+	oss << std::fixed << std::setprecision(6);
+
+	if (IsDenormalized())
+	{
+		oss << "Denormalized(" << res << ")";
+	}
+	else if (value == MAX_FLOATING_POINT_VALUE)
+	{
+		oss << "Fmax(" << res << ")";
+	}
+	else if (value == MIN_FLOATING_POINT_VALUE)
+	{
+		oss << "-Fmax(" << res << ")";
+	}
+	else if (value == POSITIVE_INFINITY_VALUE)
+	{
+		oss << "Inf(" << res << ")";
+	}
+	else if (value == NEGATIVE_INFINITY_VALUE)
+	{
+		oss << "-Inf(" << res << ")";
+	}
+	else
+	{
+		oss << "Ps2Float(" << res << ")";
+	}
+
+	return oss.str();
+}
+
+Ps2Float Ps2Float::DoAdd(Ps2Float other)
+{
+	const uint8_t roundingMultiplier = 6;
+
+	uint8_t selfExponent = Exponent;
+	int32_t resExponent = selfExponent - other.Exponent;
+
+	if (resExponent < 0)
+		return other.DoAdd(*this);
+	else if (resExponent >= 25)
+		return *this;
+
+	// http://graphics.stanford.edu/~seander/bithacks.html#ConditionalNegate
+	uint32_t sign1 = (uint32_t)((int32_t)AsUInt32() >> 31);
+	int32_t selfMantissa = (int32_t)(((Mantissa | 0x800000) ^ sign1) - sign1);
+	uint32_t sign2 = (uint32_t)((int32_t)other.AsUInt32() >> 31);
+	int32_t otherMantissa = (int32_t)(((other.Mantissa | 0x800000) ^ sign2) - sign2);
+
+	// PS2 multiply by 2 before doing the Math here.
+	int32_t man = (selfMantissa << roundingMultiplier) + ((otherMantissa << roundingMultiplier) >> resExponent);
+	int32_t absMan = abs(man);
+	if (absMan == 0)
+		return Ps2Float(0);
+
+	// Remove from exponent the PS2 Multiplier value.
+	int32_t rawExp = selfExponent - roundingMultiplier;
+
+	int32_t amount = normalizeAmounts[clz(absMan)];
+	rawExp -= amount;
+	absMan <<= amount;
+
+	int32_t msbIndex = BitScanReverse8(absMan >> 23);
+	rawExp += msbIndex;
+	absMan >>= msbIndex;
+
+	if (rawExp > 255)
+		return man < 0 ? Min() : Max();
+	else if (rawExp <= 0)
+		return Ps2Float(man < 0, 0, 0);
+
+	return Ps2Float((uint32_t)man & SIGNMASK | (uint32_t)rawExp << 23 | ((uint32_t)absMan & 0x7FFFFF));
```

Comment:
```suggestion
	return Ps2Float(((uint32_t)man & SIGNMASK) | (uint32_t)rawExp << 23 | ((uint32_t)absMan & 0x7FFFFF));
```
Should address MSVC warning -Wbitwise-op-parentheses